### ISSUE TITLE
Unused variable false negative

### ISSUE THREAD

**Issue by altoorodevelead** [PyCQA/pylint#3854]
### Steps to reproduce
1. Use this code file(test.py):
```python
print("Hello world");
A = 10
B = 15
C = ["a", "b", "c"]

def test_func():
    UNUSED_VAR = 125
    return 0
```
2. With this command:
```bash
pylint -f json > output.json test.py
```

### Current behavior
```json
[
    {
        "message": "Unnecessary semicolon", 
        "obj": "", 
        "column": 0, 
        "path": "test.py", 
        "line": 1, 
        "message-id": "W0301", 
        "type": "warning", 
        "symbol": "unnecessary-semicolon", 
        "module": "test"
    }, 
    {
        "message": "Missing module docstring", 
        "obj": "", 
        "column": 0, 
        "path": "test.py", 
        "line": 1, 
        "message-id": "C0111", 
        "type": "convention", 
        "symbol": "missing-docstring", 
        "module": "test"
    }, 
    {
        "message": "Missing function docstring", 
        "obj": "test_func", 
        "column": 0, 
        "path": "test.py", 
        "line": 6, 
        "message-id": "C0111", 
        "type": "convention", 
        "symbol": "missing-docstring", 
        "module": "test"
    }, 
    {
        "message": "Variable name \"UNUSED_VAR\" doesn't conform to snake_case naming style", 
        "obj": "test_func", 
        "column": 4, 
        "path": "test.py", 
        "line": 7, 
        "message-id": "C0103", 
        "type": "convention", 
        "symbol": "invalid-name", 
        "module": "test"
    }, 
    {
        "message": "Unused variable 'UNUSED_VAR'", 
        "obj": "test_func", 
        "column": 4, 
        "path": "test.py", 
        "line": 7, 
        "message-id": "W0612", 
        "type": "warning", 
        "symbol": "unused-variable", 
        "module": "test"
    }
]```

### Expected behavior
```json
[
    {
        "message": "Unnecessary semicolon", 
        "obj": "", 
        "column": 0, 
        "path": "test.py", 
        "line": 1, 
        "message-id": "W0301", 
        "type": "warning", 
        "symbol": "unnecessary-semicolon", 
        "module": "test"
    }, 
    {
        "message": "Missing module docstring", 
        "obj": "", 
        "column": 0, 
        "path": "test.py", 
        "line": 1, 
        "message-id": "C0111", 
        "type": "convention", 
        "symbol": "missing-docstring", 
        "module": "test"
    }, 
    {
        "message": "Missing function docstring", 
        "obj": "test_func", 
        "column": 0, 
        "path": "test.py", 
        "line": 6, 
        "message-id": "C0111", 
        "type": "convention", 
        "symbol": "missing-docstring", 
        "module": "test"
    }, 
    {
        "message": "Variable name \"UNUSED_VAR\" doesn't conform to snake_case naming style", 
        "obj": "test_func", 
        "column": 4, 
        "path": "test.py", 
        "line": 7, 
        "message-id": "C0103", 
        "type": "convention", 
        "symbol": "invalid-name", 
        "module": "test"
    }, 
    {
        "message": "Unused variable 'UNUSED_VAR'", 
        "obj": "test_func", 
        "column": 4, 
        "path": "test.py", 
        "line": 7, 
        "message-id": "W0612", 
        "type": "warning", 
        "symbol": "unused-variable", 
        "module": "test"
    }
    , 
    {
        "message": "Unused variable 'A'", 
        "obj": "", 
        "column": 1, 
        "path": "test.py", 
        "line": 2, 
        "message-id": "W0612", 
        "type": "warning", 
        "symbol": "unused-variable", 
        "module": "test"
    }
    , 
    {
        "message": "Unused variable 'B'", 
        "obj": "", 
        "column": 1, 
        "path": "test.py", 
        "line": 3, 
        "message-id": "W0612", 
        "type": "warning", 
        "symbol": "unused-variable", 
        "module": "test"
    }
    , 
    {
        "message": "Unused variable 'C'", 
        "obj": "", 
        "column": 1, 
        "path": "test.py", 
        "line": 4, 
        "message-id": "W0612", 
        "type": "warning", 
        "symbol": "unused-variable", 
        "module": "test"
    }
]
```

### pylint --version output
Same problem occurs with Python3 and Pylint3 as well.
```text
No config file found, using default configuration
pylint 1.9.5, 
astroid 1.6.6
Python 2.7.17 (default, Jul 20 2020, 15:37:01) 
[GCC 7.5.0]
```

**Comment by dbaty** [comment]
I find the JSON output hard to read when looking for differences between current and expected behaviours. If I understand correctly, a smaller, clearer example boils down to this:

```python
# pylint: disable=missing-module-docstring,missing-function-docstring,invalid-name

A = 1

def func():
    B_INSIDE_FUNC = 1
    return 2
```

The output is:

```text
$ pylint test.py 
************* Module test
test.py:6:4: W0612: Unused variable 'B_INSIDE_FUNC' (unused-variable)
```

And you are saying that pylint should complain about both `A` and `B_INSIDE_FUNC`.

pylint does not complain about `A` by default because it's a module-level variable. If you want such warnings, you need to disable the  `allow-global-unused-variables` option (which is enabled by default):

```text
$ pylint --allow-global-unused-variables=no test.py
************* Module test
test.py:6:4: W0612: Unused variable 'B_INSIDE_FUNC' (unused-variable)
test.py:3:0: W0612: Unused variable 'A' (unused-variable)
test.py:5:0: W0612: Unused variable 'func' (unused-variable)
```

This is a global option that hence cannot be enabled/disabled per module. I am not sure there are many use cases where you could enable it globally. For what's worth, [Vulture](https://pypi.org/project/vulture/) detects unused module-level variables, and there are probably other similar tools.

**Comment by hippo91** [comment]
@altoorodevelead thanks for the report.
@dbaty , one more time, thanks a lot for your investigations!
As it seems there is no bug, i close this issue. Feel free to reopen it if i am wrong.

**Comment by altoorodevelead** [comment]
Thank you so much!
That is exactly what I was looking for.
Sorry for any inconvenience.
